### PR TITLE
LLT-3629 add natlab ipv6 support

### PR DESCRIPTION
**Pull request by matislovas**
Adding dual stack configuration to nat-lab.

### :ballot_box_with_check: Definition of Done checklist
- [x] Commit history is clean
- [x] README.md is updated
- [x] changelog.md is updated

### CODE REVIEW

File: nat-lab/docker-compose.yml [hunk 1]
```diff
@@ -144,6 +146,28 @@ services:
     volumes:
       - ../../:/libtelio-build
 
+  # Create one open internet client for
+  # cryptography tests
+  open-internet-client-dual-stack:
```
[the remark this answers is not in the document]
Comment (matislovas) [reply nat-lab/docker-compose.yml:151]: The idea was, that `open-internet-client-0x` are _ordinary_ IPv4-only clients, while this fella would be a dual-stack (to test IPv6 exit node functionality).

File: nat-lab/run_local.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 
 PROJECT_ROOT = os.path.normpath(os.path.dirname(os.path.realpath(__file__)) + "/../..")
 
-TEST_TIMEOUT = 60
+TEST_TIMEOUT = 80
```
[the remark this answers is not in the document]
Comment (matislovas) [reply nat-lab/run_local.py:11]: Yeah, I ran tests multiple times with dual stack configuration, and it crashed on ~97% because of the timeout ...

File: nat-lab/tests/test_connection_states.py [hunk 3]
```diff
@@ -31,10 +31,7 @@
         pytest.param(
             ConnectionTag.WINDOWS_VM,
             telio.AdapterType.WindowsNativeWg,
-            marks=[
-                pytest.mark.windows,
-                pytest.mark.xfail(reason="test is flaky - Jira issue: LLT-4064"),
```
[the remark this answers is not in the document]
Comment (matislovas) [reply nat-lab/tests/test_connection_states.py:36]: It is not reproducible in `WindowsNativeWg` scenario, so that's why I unmarked it (only the `WindowsWgGo` is problematic).

File: nat-lab/tests/utils/linux_router.py [hunk 4]
```diff
@@ -173,15 +196,85 @@ async def delete_exit_node_route(self) -> None:
...
     async def disable_path(self, address: str) -> AsyncIterator:
+        addr_proto = self.check_ip_address(address)
+
+        if addr_proto == None:
+            return
+
         await self._connection.create_process(
             [
-                "iptables",
+                ("ip" if addr_proto == IPProto.IPv4 else "ip6") + "tables",
```
[the remark this answers is not in the document]
Comment (matislovas) [reply nat-lab/tests/utils/linux_router.py]: Fixed

File: nat-lab/tests/utils/linux_router.py [hunk 5]
```diff
@@ -24,41 +24,79 @@ class LinuxRouter(Router):
...
+                    "-4",
+                    "route",
+                    "add",
+                    "100.64.0.0/10",
+                    "dev",
+                    self._interface_name,
+                ],
+            ).execute()
+
+        if self.ip_stack == IPStack.IPv6 or self.ip_stack == IPStack.IPv4v6:
```
[the remark this answers is not in the document]
Comment (matislovas) [reply nat-lab/tests/utils/linux_router.py]: Fixed

File: nat-lab/tests/utils/mac_router.py [hunk 6]
```diff
@@ -16,54 +15,82 @@ class MacRouter(Router):
...
+                        "ifconfig",
+                        self._interface_name,
+                        "inet6",
+                        "add",
+                        address,
+                        "prefixlen",
+                        "64",
+                    ],
+                ).execute()
+            else:
```
[the remark this answers is not in the document]
Comment (matislovas) [reply nat-lab/tests/utils/mac_router.py]: Fixed

File: nat-lab/tests/utils/linux_router.py [hunk 7]
```diff
@@ -24,41 +24,79 @@ class LinuxRouter(Router):
...
-            [
-                "ip",
-                "route",
-                "add",
-                "100.64.0.0/10",
-                "dev",
-                self._interface_name,
-            ],
-        ).execute()
+        if self.ip_stack == IPStack.IPv4 or self.ip_stack == IPStack.IPv4v6:
```
[the remark this answers is not in the document]
Comment (matislovas) [reply nat-lab/tests/utils/linux_router.py]: Fixed

File: nat-lab/tests/utils/linux_router.py [hunk 8]
```diff
@@ -237,9 +331,14 @@ async def disable_path(self, address: str) -> AsyncIterator:
...
     async def break_tcp_conn_to_host(self, address: str) -> AsyncIterator:
+        addr_proto = self.check_ip_address(address)
+
+        if addr_proto == None:
+            return
+
         await self._connection.create_process(
             [
-                "iptables",
+                ("ip" if addr_proto == IPProto.IPv4 else "ip6") + "tables",
```
[the remark this answers is not in the document]
Comment (matislovas) [reply nat-lab/tests/utils/linux_router.py]: Fixed

File: nat-lab/tests/utils/windows_router.py [hunk 9]
```diff
@@ -1,53 +1,93 @@
...
+                        "netsh",
+                        "interface",
+                        "ipv6",
+                        "add",
+                        "address",
+                        self._interface_name,
+                        address + "/128",
+                    ]
+                ).execute()
+            else:
```
[the remark this answers is not in the document]
Comment (matislovas) [reply nat-lab/tests/utils/windows_router.py]: Fixed

File: nat-lab/tests/utils/linux_router.py [hunk 10]
```diff
@@ -24,41 +24,79 @@ class LinuxRouter(Router):
...
+                        "ip",
+                        "-6",
+                        "addr",
+                        "add",
+                        address,
+                        "dev",
+                        self._interface_name,
+                    ],
+                ).execute()
+            else:
```
[the remark this answers is not in the document]
Comment (matislovas) [reply nat-lab/tests/utils/linux_router.py]: Fixed

File: nat-lab/tests/utils/mac_router.py [hunk 11]
```diff
@@ -16,54 +15,82 @@ class MacRouter(Router):
...
+            if addr_proto == IPProto.IPv4:
+                await self._connection.create_process(
+                    [
+                        "ifconfig",
+                        self._interface_name,
+                        "inet",
+                        "add",
+                        address,
+                        "255.192.0.0",
+                        address,
```
[the remark this answers is not in the document]
Comment (matislovas) [reply nat-lab/tests/utils/mac_router.py:38]: Yeah that looked strange to me as well, but it was implemented the same way before and it worked. But will change that and will see how it works ...

File: nat-lab/tests/utils/linux_router.py [hunk 12]
```diff
@@ -138,31 +162,13 @@ async def delete_vpn_route(self):
...
-    async def delete_exit_node_route(self) -> None:
-        try:
+        if self.ip_stack in [IPStack.IPv4, IPStack.IPv4v6]:
             await self._connection.create_process(
                 [
                     "iptables",
                     "-t",
                     "nat",
-                    "-D",
+                    "-A",
```
Comment (Hasan6979) [inline nat-lab/tests/utils/linux_router.py:171]: Why is this changed from D to A here ?

**Review by Hasan6979 — COMMENTED** [review]
+ 1.0

File: nat-lab/tests/utils/linux_router.py (hunk 12 above)
Comment (matislovas) [reply nat-lab/tests/utils/linux_router.py:171]: The `git` changes are just projected weirdly here. If you'll check the functions, `create_exit_node_route` has `-A`, and `delete_exit_node_route` has `-D` options.

File: nat-lab/docker-compose.yml [hunk 13]
```diff
@@ -534,6 +559,15 @@ networks:
       config:
         - subnet: 10.0.0.0/16
 
+  # Network representing internet with global ipv6 addresses
+  internet_v6:
```
[the remark this answers is not in the document]
Comment (matislovas) [reply nat-lab/docker-compose.yml:563]: Emm, for the sake of legacy/simplicity I would leave the rest with `internet`, unless you strongly object :D

File: nat-lab/tests/test_connection_states.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gytsto) [reply nat-lab/tests/test_connection_states.py:31]: Is it legit `WindowsNativeWg` problem? or you just managed to reporduce on it?

File: nat-lab/docker-compose.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jauler) [reply nat-lab/docker-compose.yml:151]: We do have some tests planned for IPv4-only, so separating them out into separate container allows us to choose which client we would like to use in the test.

File: nat-lab/bin/vpn-server [hunk 14]
```diff
@@ -2,10 +2,16 @@
...
+sysctl -w net.ipv6.conf.all.disable_ipv6=0
+sysctl -w net.ipv6.conf.default.disable_ipv6=0
+
 /opt/bin/wireguard-go wg0
 wg setconf wg0 /wg0.conf
 
-ip addr add dev wg0 100.64.0.1/10
+ip -4 addr add dev wg0 100.64.0.1/10
+# TODO correct subnet when we'll decide about the range
+ip -6 addr add dev wg0 fd00::1/64
```
Comment (Jauler) [inline nat-lab/bin/vpn-server:14]: Is this the correct range? In teh nat-lab/data/wg0.conf it seems like the range ips assigned are `fd00::dead:x/128` and `fd00::1/64` does not include those.

File: nat-lab/docker-compose.yml [hunk 15]
```diff
@@ -158,6 +182,7 @@ services:
       CLIENT_GATEWAY_PRIMARY: 192.168.101.254
     extra_hosts:
       - "photo-album:10.0.80.80"
+      - "photo-album:2001:0db8:85a4::adda:edde:0005"
```
Comment (Jauler) [inline nat-lab/docker-compose.yml:185]: is this correct? IPv6 address of a photo album as `extra_hosts` should be added to all containers which has IPv6 connectivity (e.g. `nat-lab-open-internet-client-dual-stack-1`) the con-client-01 -> not sure if this is relevant.

**Review by Jauler — COMMENTED** [review]
Could you add one test case in `test_direct_connection.py` with dual stack container, e.g. open_internet and dual_stack into `UHP_conn_client_types`?

**Comment by matislovas** [comment]
> Could you add one test case in `test_direct_connection.py` with dual stack container, e.g. open_internet and dual_stack into `UHP_conn_client_types`?

I thought that this PR should only pave the way for future tests, and I will actually add all IPv6 related tests in [WG-IPv6 PR](https://github.com/NordSecurity/libtelio/pull/61).

File: nat-lab/docker-compose.yml (hunk 15 above)
Comment (matislovas) [reply nat-lab/docker-compose.yml:185]: Good catch, fixed!

File: nat-lab/bin/vpn-server (hunk 14 above)
Comment (Jauler) [reply nat-lab/bin/vpn-server:14]: oh damn :D 

Sorry, I misread the IP address due to double colon.

File: nat-lab/docker-compose.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jauler) [reply nat-lab/docker-compose.yml:151]: I think this discussion can be marked as resolved

File: nat-lab/tests/test_connection_states.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jauler) [reply nat-lab/tests/test_connection_states.py:31]: I think this discussion can be marked as resolved

**Comment by matislovas** [comment]
> > > Could you add one test case in `test_direct_connection.py` with dual stack container, e.g. open_internet and dual_stack into `UHP_conn_client_types`?
> > 
> > 
> > I thought that this PR should only pave the way for future tests, and I will actually add all IPv6 related tests in [WG-IPv6 PR](https://github.com/NordSecurity/libtelio/pull/61).
> 
> That one test case is kinda intended to check whether we have not missed something within the setup itself ;). It is ok, to have a separate PR for WG-IPv6 tests ;)

Addressed!

**Review by gytsto — APPROVED** [review]

**Review by Jauler — APPROVED** [review]
+1.5
